09-schema/03-rules.md

```diff
@@ -274,6 +274,112 @@ TypeQLDefine query = TypeQL.define(
+```
+p--q, q--r, r--s, s--t  (already in the database)
+
+p--r, q--s, r--t,       (Inferred)
+p--s, q--t, p--t        (Inferred)                    
+```
+
+Concretely, We would generate an `edge` relation _for every pair_ of nodes reachable from `p` - a **quadratic** number of relations.
+
+ The following section describes a recipe to answer forward transitivity queries materialising only a **linear** number of relations. Later, the recipe is extended to backward queries and undirected relations.
+
+### Forward transitivity
+We first define separate types for the persisted and (inferred) transitive version of the relation.
+For the example above, we use `edge` as the base relation type and `forward-reachable` as the inferred relation. We then update the rule as follows: 
+```typeql
+define
+rule forward-transitivity-base:
+when{
+    (from: $x, to: $y) isa edge;
+} then {
+    (from: $x, to: $y) isa forward-reachable;
+};
+
+rule forward-transitivity-recursive:
+when{
+    (from: $x, to: $y) isa forward-reachable;
+    (from: $y, to: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa forward-reachable;
+};
+```
+
+We can intepret this as finding a path and extend it by one. Querying all nodes reachable from p in the chain p-q-r-s-t would generate the following relations:
+```
+p-q, q-r, r-s, s-t      (edges already in the database)
+
+p--q,                   (Inferred with the first rule)
+p--r,                   (Inferred with the second rule)
+p--s,                   (Inferred with the second rule)
+p--t                    (Inferred with the second rule)
+```
+Here, we only generate one relation for _**each node**_ reachable from p, bringing the complexity down from quadratic in to linear in the number of reachable nodes.
+
+### Backward transitivity
+To see what happens when we try to compute backwards transitivity using the above formulation, consider the query to find all nodes from which `t` is reachable in the same chain `p-q-r-s-t`. The second rule is now executed backwards - first checking all nodes `$y` from which there is an edge to `t`. Then it recursively queries all nodes reachable from `$y`. Thus, a relation is generated for every pair of nodes which are reachable from `t`.
+
+To answer backward transitive queries, we simply need a backwards version of the transitive relation and rules. Intuitively, This approach computes forward-transitivity on the reversed graph.
+```typeql
+define
+rule backward-transitivity-base:
+when{
+    (to: $x, from: $y) isa edge;
+} then {
+    (to: $x, from: $y) isa backward-reachable;
+};
+
+rule backward-transitivity-recursive:
+when{
+    (to: $x, from: $y) isa backward-reachable;
+    (to: $y, from: $z) isa edge;
+} then {
+    (to: $x, from: $z) isa backward-reachable;
+};
+```
+
+### Undirected transitivity
+We can use the same formulation for undirected graphs. If the undirected edges are defined by the relations `(node: $x, node: $y) isa edge;` then the rules would read:
```

Comment:
Sometimes, the original dataset is undirected and we want to be able to get all concepts reachable in a either direction from a fixed starting concept. 

We can apply a trick to convert an undirected reachabilty into a directed transitivity...